### PR TITLE
Adjoint matrix and some jacobians

### PR DESCRIPTION
**Pull request by mateuszbaran**
A generic implementation of adjoint matrix with optimized implementations for selected groups. I've adjusted optimized formulas to match our conventions so it's not exactly the same as in original papers for SE(2) and SE(3).

### CODE REVIEW

**Comment by olivierverdier** [comment]
I think this is what I already implemented here (in a more general setting): https://olivierverdier.github.io/ManifoldGroupUtils.jl/stable/#ManifoldGroupUtils.matrix_from_lin_endomorphism-Tuple{Any,%20Any,%20ManifoldsBase.AbstractBasis}

**Comment by mateuszbaran** [comment]
That could be true but I'm not sure how to use your function to get the result that I want.

**Comment by olivierverdier** [comment]
If you have a vector $`X`$ in the Lie algebra, I think that it would be `matrix_from_lin_endomorphism(G, Y -> lie_bracket(G, X, Y), DefaultOrthogonalBasis())`.

**Comment by mateuszbaran** [comment]
I'm not sure it's the same adjoint. Anyway, here the primary need is to be able to overload the computation for selected manifolds with optimized code and doing that with `matrix_from_lin_endomorphism` would be tricky.

File: src/groups/special_euclidean.jl [hunk 1]
```diff
@@ -190,6 +190,61 @@ function adjoint_action(
...
+        R[1, 1] R[1, 2] t[2]/sqrt(2)
+        R[2, 1] R[2, 2] -t[1]/sqrt(2)
+        0 0 1
+    ]
+end
+@doc raw"""
+    adjoint_matrix(::SpecialEuclidean{TypeParameter{Tuple{3}}}, p)
+
+Compute the adjoint matrix for the group [`SpecialEuclidean`](@ref)`(3)` at point `p`
+in default coordinates. The formula follows Section 10.6.9 in [Chirikjian:2012] with
```
Comment (kellertuer) [inline src/groups/special_euclidean.jl:222]:
```suggestion
in default coordinates. The formula follows Section 10.6.9 in [Chirikjian:2012](@cite) with
```

File: src/groups/special_euclidean.jl [hunk 2]
```diff
@@ -190,6 +190,61 @@ function adjoint_action(
     return TFVector([cross(t, Rω) + R * r; Rω], fX.basis)
 end
 
+@doc raw"""
+    adjoint_matrix(::SpecialEuclidean{TypeParameter{Tuple{2}}}, p)
+
+Compute the adjoint matrix for the group [`SpecialEuclidean`](@ref)`(2)` at point `p`
+in default coordinates. The formula follows Section 10.6.2 in [Chirikjian:2012]
```
Comment (kellertuer) [inline src/groups/special_euclidean.jl:197]:
```suggestion
in default coordinates. The formula follows Section 10.6.2 in [Chirikjian:2012](@cite)
```

File: src/groups/special_euclidean.jl [hunk 3]
```diff
@@ -494,6 +549,128 @@ function exp_lie!(G::SpecialEuclidean{TypeParameter{Tuple{3}}}, q, X)
...
+    jacobian_exp_inv_argument(
+        M::SpecialEuclidean{TypeParameter{Tuple{2}}},
+        p,
+        X,
+    )
+
+Compute Jacobian matrix of the invariant exponential map on [`SpecialEuclidean`](@ref)`(2)`.
+The formula reads
+````math
+\begin{pmatrix}
```
Comment (kellertuer) [inline src/groups/special_euclidean.jl:562]: one could maybe do the fractions as ``\frac{1}{\theta}(\sin(\theta)`` ? 

And there are a few `\theta`s that could be turned into their UTF8s like the other are.

File: src/groups/special_euclidean.jl [hunk 4]
```diff
@@ -494,6 +549,128 @@ function exp_lie!(G::SpecialEuclidean{TypeParameter{Tuple{3}}}, q, X)
...
+R & Q \\
+0_{3×3} & R
+\end{pmatrix},
+````
+where ``R`` is the Jacobian of exponential map on [`Rotations`](@ref)`(3)` with respect to
+the argument, and ``Q`` is
+````math
+\frac{1}{2} T - \frac{θ - \sin(θ)}{θ^3} (X_r T + T X_r + X_r T X_r) + \frac{1 - \frac{θ^2}{2} - \cos(θ)}{θ^4} (X_r^2 T + T X_r^2 - 3 X_r T X_r) + \frac{1}{2}\left(\frac{1 - \frac{θ^2}{2} - \cos(θ)}{θ^4} - 3 \frac{θ - \sin(θ) - \frac{θ^3}{6}}{θ^5}\right) (X_r T X_r^2 + X_r^2 T X_r)
```
Comment (kellertuer) [inline src/groups/special_euclidean.jl:591]:
```suggestion
Q = \frac{1}{2} T - \frac{θ - \sin(θ)}{θ^3} (X_r T + T X_r + X_r T X_r) + \frac{1 - \frac{θ^2}{2} - \cos(θ)}{θ^4} (X_r^2 T + T X_r^2 - 3 X_r T X_r) + \frac{1}{2}\left(\frac{1 - \frac{θ^2}{2} - \cos(θ)}{θ^4} - 3 \frac{θ - \sin(θ) - \frac{θ^3}{6}}{θ^5}\right) (X_r T X_r^2 + X_r^2 T X_r)
```

and maybe we cloud split this in a few lines? This is very long.

File: src/groups/special_orthogonal.jl [hunk 5]
```diff
@@ -1,17 +1,35 @@
...
 
 function SpecialOrthogonal(n; parameter::Symbol=:type)
     return GeneralUnitaryMultiplicationGroup(Rotations(n; parameter=parameter))
 end
 
+"""
+    adjoint_matrix(::SpecialOrthogonal{TypeParameter{Tuple{2}}}, p)
+
+Compte the adjoint matrix for [`SpecialOrthogonal`](@ref)`(2)` at point `p`, which is equal
+to `1`. See [SolaDerayAtchuthan:2021], Appendix A.
```
Comment (kellertuer) [inline src/groups/special_orthogonal.jl:19]:
```suggestion
to `1`. See [SolaDerayAtchuthan:2021](@cite), Appendix A.
```

**Review by kellertuer — COMMENTED** [review]
Here is a few small comments.

File: src/groups/special_euclidean.jl (hunk 4 above)
Comment (mateuszbaran) [reply src/groups/special_euclidean.jl]: Sure, I can do that.

File: src/groups/special_euclidean.jl (hunk 3 above)
Comment (mateuszbaran) [reply src/groups/special_euclidean.jl:562]: Something like in the commit I've just made?

Comment (kellertuer) [reply src/groups/special_euclidean.jl:562]: Perfect.

**Review by kellertuer — APPROVED** [review]

**Pull request merged**